source/secure_tunneling_operations.c

```diff
@@ -0,0 +1,670 @@
+        (void *)message_view,
+        message_view->payload.len);
+}
+
+static size_t s_aws_secure_tunnel_message_compute_storage_size(
+    const struct aws_secure_tunnel_message_view *message_view) {
+    size_t storage_size = message_view->payload.len;
+    storage_size += message_view->service_id.len;
+
+    return storage_size;
+}
+
+int aws_secure_tunnel_message_storage_init(
+    struct aws_secure_tunnel_message_storage *message_storage,
+    struct aws_allocator *allocator,
+    const struct aws_secure_tunnel_message_view *message_options,
+    enum aws_secure_tunnel_operation_type type) {
+
+    AWS_ZERO_STRUCT(*message_storage);
+    size_t storage_capacity = s_aws_secure_tunnel_message_compute_storage_size(message_options);
+    if (aws_byte_buf_init(&message_storage->storage, allocator, storage_capacity)) {
+        return AWS_OP_ERR;
+    }
+
+    struct aws_secure_tunnel_message_view *storage_view = &message_storage->storage_view;
+
+    storage_view->type = message_options->type;
+    storage_view->ignorable = message_options->ignorable;
+    storage_view->stream_id = message_options->stream_id;
+
+    switch (type) {
+        case AWS_STOT_DATA:
+            storage_view->type = AWS_SECURE_TUNNEL_MT_DATA;
+            break;
+        case AWS_STOT_STREAM_START:
+            storage_view->type = AWS_SECURE_TUNNEL_MT_STREAM_START;
+            break;
+        case AWS_STOT_STREAM_RESET:
+            storage_view->type = AWS_SECURE_TUNNEL_MT_STREAM_RESET;
+            break;
+        default:
+            storage_view->type = AWS_SECURE_TUNNEL_MT_UNKNOWN;
+            break;
+    }
+
+    storage_view->service_id = message_options->service_id;
+    if (aws_byte_buf_append_and_update(&message_storage->storage, &storage_view->service_id)) {
+        return AWS_OP_ERR;
+    }
+
+    storage_view->payload = message_options->payload;
+    if (aws_byte_buf_append_and_update(&message_storage->storage, &storage_view->payload)) {
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+void aws_secure_tunnel_message_storage_clean_up(struct aws_secure_tunnel_message_storage *message_storage) {
+    aws_byte_buf_clean_up(&message_storage->storage);
+}
+
+/* Sets the stream id on outbound message based on the service id (or lack of for V1) to the current one being used. */
+static int s_aws_secure_tunnel_operation_message_set_stream_id(
+    struct aws_secure_tunnel_operation *operation,
+    struct aws_secure_tunnel *secure_tunnel) {
+
+    struct aws_secure_tunnel_operation_message *message_op = operation->impl;
+    int32_t stream_id = INVALID_STREAM_ID;
+
+    struct aws_secure_tunnel_message_view *message_view = &message_op->options_storage.storage_view;
+
+    if (message_view->service_id.len > 0) {
+        struct aws_string *service_id = NULL;
+        service_id = aws_string_new_from_cursor(secure_tunnel->allocator, &message_view->service_id);
+
+        if (secure_tunnel->config->service_id_1 != NULL &&
```

Comment:
might be simpler and more future-proof to have the service ids to be tracked in a hash table.  It's tricky because you'll need to create a structure ala:
```
struct service_id_entry {
   struct aws_byte_cursor service_id_cursor;
   struct aws_byte_buf service_id;
   int id;
};

```

The struct is what goes in the value and you use cursors (including the embedded one when adding) for lookup.  That's what we did with topic aliasing.

Reply:
I'll revisit using a hash table. I originally started setting up a hash table but ended up swapping to this because we were always guaranteed to only have up to three service ids. I think some things changed after I set this up as well in terms of what the secure tunnel service was providing vs. what I thought we had to provide upon secure tunnel creation. A hash table would def simplify a number of areas related to getting/setting/modifying the stream id.

Reply:
Could we at least use a fixed size array then?

Reply:
Converted to a hash table.